### PR TITLE
Generalize identity determination in FE_SimplexP

### PR DESCRIPTION
**Pull request by peterrum**
@sebproell Maybe you could use this as starting point.

### CODE REVIEW

File: source/fe/fe_simplex_p.cc [hunk 1]
```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
...
-      // after the 3 vertex points in the complete list of unit support points
+      std::vector<std::pair<unsigned int, unsigned int>> identities;
+
+      for (unsigned int i = 0; i < this->degree - 1; ++i)
+        for (unsigned int j = 0; j < fe_p_other->degree - 1; ++j)
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->unit_support_points
+                          [i + this->reference_cell().n_vertices()][0]) < 1e-14)
+            identities.emplace_back(i, j);
```
Comment (bangerth) [inline source/fe/fe_simplex_p.cc:775]: This looks reasonable, but I think that the comment actually was helpful in understanding the code. Why did you remove it?

Separately, do we fill `face_unit_support_points` for `FE_SimplexP`? If so, it might be easier to use because then the indexing becomes simpler.

File: source/fe/fe_simplex_p.cc [hunk 2]
```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
...
       std::vector<std::pair<unsigned int, unsigned int>> identities;
 
       for (unsigned int i = 0; i < this->degree - 1; ++i)
         for (unsigned int j = 0; j < fe_p_other->degree - 1; ++j)
-          if (std::fabs(this->unit_support_points[i + 3][0] -
-                        fe_p_other->unit_support_points[i + 3][0]) < 1e-14)
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->get_unit_support_points()
+                          [i + this->reference_cell().n_vertices()][1]) < 1e-14)
```
Comment (bangerth) [inline source/fe/fe_simplex_p.cc:789]: This block could use a comment that says that if this function is called with both an `FE_SimplexP` and a `FE_PyramidP`, then we are clearly at a triangular face of the pyramid and can treat it the same way we did above for two `FE_SimplexP` meeting.

File: source/fe/fe_simplex_p.cc [hunk 3]
```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
...
             identities.emplace_back(i, j);
-          else
-            {
-              // If nodes are not located in the same place, we have to
-              // interpolate. This is then not handled through the
-              // current function, but via interpolation matrices that
-              // result in constraints, rather than identities. Since
-              // that happens in a different function, there is nothing
-              // for us to do here.
-            }
```
Comment (bangerth) [inline source/fe/fe_simplex_p.cc:786]: I added this in #14444 because I was wondering how this is working if the `if` above is not satisfied. I think the comment is still relevant and if you don't mind too much, please leave it here and duplicate it in the `FE_SimplexP` case above.

File: source/fe/fe_simplex_p.cc [hunk 4]
```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
...
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->unit_support_points
+                          [i + this->reference_cell().n_vertices()][0]) < 1e-14)
+            identities.emplace_back(i, j);
 
+      return identities;
+    }
+  else if (const auto *fe_p_other =
+             dynamic_cast<const FE_PyramidP<dim, spacedim> *>(&fe_other))
```
Comment (bangerth) [inline source/fe/fe_simplex_p.cc:780]: I believe that you can avoid the code duplication in the same way as you did for the vertex case above: Just have a sequence of `dynamic_cast<...> || dynamic_cast<...>` questions in the `if`, and then access the information via `fe_other` rather than via `fe_p_other`. I believe that nothing you do in the `if` block actually requires access to information of the derived object.

**Review by bangerth — COMMENTED** [review]
Generally ok, but I have a couple of comments.

**Comment by bangerth** [comment]
@peterrum If you address the comments, I will gladly look at it again!

**Comment by peterrum** [comment]
We can finally can close this. The content has been moved to https://github.com/dealii/dealii/pull/17576#pullrequestreview-2248329694. Thanks @dominiktassilostill!
